api/src/main/java/org/openmrs/api/db/hibernate/HibernateProviderDAO.java

```diff
@@ -136,20 +145,28 @@ public ProviderAttribute getProviderAttributeByUuid(String uuid) {
 	@Override
 	public List<Provider> getProviders(String name, Map<ProviderAttributeType, String> serializedAttributeValues,
 	        Integer start, Integer length, boolean includeRetired) {
-		Criteria criteria = prepareProviderCriteria(name, includeRetired);
-		if (start != null) {
-			criteria.setFirstResult(start);
-		}
-		if (length != null) {
-			criteria.setMaxResults(length);
-		}
+		Session session = sessionFactory.getCurrentSession();
+		CriteriaBuilder cb = session.getCriteriaBuilder();
+		CriteriaQuery<Provider> cq = cb.createQuery(Provider.class);
+		Root<Provider> root = cq.from(Provider.class);
+
+		List<Predicate> predicates = prepareProviderCriteria(cb, root, name, includeRetired);
```

Comment:
Is there an easy way of doing `distinct(true)` from within `prepareProviderCriteria`?

Reply:
I don't think there is while keeping the method flexible enough for reuse. In some of the first migrations I tried doing it by passing the CriteriaQuery object into the method to allow me to do this, but then when we want to do certain things outside the method like counts it doesn't work because the query object specifies the return type of the result

Reply:
OK 😊